src/include/planner/logical_plan/logical_operator/logical_aggregate.h

```diff
@@ -21,30 +28,40 @@ class LogicalAggregate : public LogicalOperator {
 
     std::string getExpressionsForPrinting() const override;
 
-    inline bool hasExpressionsToGroupBy() const { return !expressionsToGroupBy.empty(); }
-    inline binder::expression_vector getExpressionsToGroupBy() const {
-        return expressionsToGroupBy;
+    inline bool hasKeyExpressions() const { return !keyExpressions.empty(); }
+    inline binder::expression_vector getKeyExpressions() const { return keyExpressions; }
+    inline void setKeyExpressions(binder::expression_vector expressions) {
+        keyExpressions = std::move(expressions);
+    }
+    inline binder::expression_vector getPayloadExpressions() const { return payloadExpressions; }
+    inline void setPayloadExpressions(binder::expression_vector expressions) {
+        payloadExpressions = std::move(expressions);
     }
-    inline size_t getNumAggregateExpressions() const { return expressionsToAggregate.size(); }
-    inline std::shared_ptr<binder::Expression> getAggregateExpression(
-        common::vector_idx_t idx) const {
-        return expressionsToAggregate[idx];
+    inline binder::expression_vector getAllGroupByExpressions() const {
+        binder::expression_vector result;
+        result.insert(result.end(), keyExpressions.begin(), keyExpressions.end());
+        result.insert(result.end(), payloadExpressions.begin(), payloadExpressions.end());
+        return result;
     }
-    inline binder::expression_vector getExpressionsToAggregate() const {
-        return expressionsToAggregate;
+    inline binder::expression_vector getAggregateExpressions() const {
+        return aggregateExpressions;
     }
 
     inline std::unique_ptr<LogicalOperator> copy() override {
         return make_unique<LogicalAggregate>(
-            expressionsToGroupBy, expressionsToAggregate, children[0]->copy());
+            keyExpressions, payloadExpressions, aggregateExpressions, children[0]->copy());
     }
 
 private:
     bool hasDistinctAggregate();
+    void insertAllExpressionsToGroupAndScope(f_group_pos groupPos);
 
 private:
-    binder::expression_vector expressionsToGroupBy;
-    binder::expression_vector expressionsToAggregate;
+    binder::expression_vector keyExpressions;
+    // A payload expression depend on a keyExpression (e.g. a.age depends on a.ID) and will not be
+    // treated as a hash key during hash aggregation.
+    binder::expression_vector payloadExpressions;
```

Comment:
Your call but I'd call this dependentKeyExpressions to be more specific. `payload` refers generally to values that correspond to keys in maps (e.g., `aggregateExpressions` can also be referred to as payloads according to this terminology.  If you do the change, I'd change the PR comments and in some other places throughout the code as well.